File: Windows/FireflyDevice/FDFireflyDeviceLogger.cpp
//
//  FDFireflyDeviceLogger.cpp
//  FireflyDevice
//
//  Created by Denis Bohm on 12/21/13.
//  Copyright (c) 2013-2014 Firefly Design LLC / Denis Bohm. All rights reserved.
//

#include "FDFireflyDeviceLogger.h"
#include "FDString.h"

#include <Windows.h>

namespace FireflyDesign {

	static std::shared_ptr<FDFireflyDeviceLog> fireflyDeviceLogger;

	void FDFireflyDeviceLogger::setLog(std::shared_ptr<FDFireflyDeviceLog> log) {
		fireflyDeviceLogger = log;
	}

	std::shared_ptr<FDFireflyDeviceLog> FDFireflyDeviceLogger::getLog() {
		return fireflyDeviceLogger;
	}

	void FDFireflyDeviceLogger::log(std::shared_ptr<FDFireflyDeviceLog> log, std::string file, unsigned line, std::string method, std::string format, ...) {
		va_list args;
		va_start(args, format);
		char buffer[512];
		vsnprintf_s(buffer, sizeof(buffer), format.c_str(), args);
		va_end(args);
		std::string message(buffer);

		size_t index = file.find_last_of('\\');
		if (index != std::string::npos) {
			file = file.substr(index + 1);
		}

		if (!log) {
			log = fireflyDeviceLogger;
		}
		if (log) {
			log->log(file, line, method, message);
		} else {
			std::string s = FDString::format("%s:%lu %s %s\n", file.c_str(), (unsigned long)line, method.c_str(), message.c_str());
			printf("%s", s.c_str());
			std::wstring ws(s.begin(), s.end());
			OutputDebugString(ws.c_str());
		}
	}

}

File: FireflyDevice/FireflyDevice/FDFireflyDeviceLogger.m
//
//  FDFireflyDeviceLogger.m
//  FireflyDevice
//
//  Created by Denis Bohm on 12/21/13.
//  Copyright (c) 2013-2014 Firefly Design LLC / Denis Bohm. All rights reserved.
//

#import <FireflyDevice/FDFireflyDeviceLogger.h>

#include <string.h>

static id<FDFireflyDeviceLog> fireflyDeviceLogger;

@implementation FDFireflyDeviceLogger

+ (void)setLog:(id<FDFireflyDeviceLog>)log
{
    fireflyDeviceLogger = log;
}

+ (id<FDFireflyDeviceLog>)log
{
    return fireflyDeviceLogger;
}

+ (void)log:(id<FDFireflyDeviceLog>)log file:(char *)file line:(NSUInteger)line class:(Class)class method:(NSString *)method format:(NSString *)format, ...
{
    char *short_file = strrchr(file, '/');
    if (short_file != NULL) {
        file = short_file + 1;
    }
    
    va_list args;
    va_start(args, format);
    NSString *message = [[NSString alloc] initWithFormat:format arguments:args];
    va_end(args);
    
    if (log == nil) {
        log = fireflyDeviceLogger;
    }
    if (log != nil) {
        [log logFile:file line:line class:class method:method message:message];
    } else {
        NSLog(@"log: %s:%lu %@.%@ %@", file, (unsigned long)line, NSStringFromClass(class), method, message);
    }
}

@end
